### PR TITLE
Add multiple work item consumers

### PR DESCRIPTION
**Pull request by oleksandr-didyk**
Resolves https://github.com/dotnet/arcade-services/issues/4022

Build: https://dev.azure.com/dnceng/internal/_build/results?buildId=2560100&view=results

### CODE REVIEW

File: src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemConfiguration.cs [hunk 1]
```diff
@@ -40,7 +41,20 @@ public static void AddWorkItemQueues(this IHostApplicationBuilder builder, Defau
             builder.Configuration.GetRequiredValue(WorkItemQueueNameConfigurationKey);
         builder.Services.Configure<WorkItemConsumerOptions>(
             builder.Configuration.GetSection(WorkItemConsumerOptions.ConfigurationKey));
-        builder.Services.AddHostedService<WorkItemConsumer>();
+
+        var consumerCountConfig = builder.Configuration.GetRequiredValue(WorkItemConsumerCountConfigurationKey);
+        if (int.TryParse(consumerCountConfig, out var consumerCount))
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemConfiguration.cs:46]: I'd maybe make this just `Parse` so that it's obvious this needs to be configured. Otherwise it might be quite hard to figure out why there's no listeners.

File: src/ProductConstructionService/ProductConstructionService.Common/MetricRecorder.cs [hunk 2]
```diff
@@ -16,7 +16,7 @@ public class MetricRecorder : IMetricRecorder
     public const string PcsMetricsNamespace = "ProductConstructionService.Metrics";
     private const string WaitTimeMetricName = "pcs.queue.wait_time";
 
-    private readonly Counter<int> _queueWaitTimeCounter;
+    private Counter<int> _queueWaitTimeCounter;
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Common/MetricRecorder.cs:19]: Why this change?

File: src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs [hunk 3]
```diff
@@ -41,7 +45,11 @@ public async Task<WorkItemScope> BeginWorkItemScopeWhenReadyAsync()
 
     private async Task WorkItemFinishedAsync()
     {
-        await _state.SetStoppedIfStoppingAsync();
+        Interlocked.Decrement(ref _activeWorkItems);
+        if (_activeWorkItems == 0)
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:49]: Doesn't `Interlocked.Decrement` return the value?

File: src/ProductConstructionService/ProductConstructionService.Common/MetricRecorder.cs (hunk 2 above)
Comment (oleksandr-didyk) [reply src/ProductConstructionService/ProductConstructionService.Common/MetricRecorder.cs:19]: `Interlocked` works with `ref` and `readonly` fields cannot be used as `ref`s

File: src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs (hunk 3 above)
Comment (oleksandr-didyk) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:49]: Yeah, both `Decrement` and `Increment` return the new value of the variable.

Oh, meaning it can be used in the conditional? I just felt this way was a bit more verbose

File: src/ProductConstructionService/ProductConstructionService.Common/MetricRecorder.cs (hunk 2 above)
Comment (oleksandr-didyk) [reply src/ProductConstructionService/ProductConstructionService.Common/MetricRecorder.cs:19]: Oh, my bad, I change how it works; will change it back

File: src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs (hunk 3 above)
Comment (premun) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:49]: A bit more verbose but also not atomic anymore? :)

Comment (oleksandr-didyk) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:49]: That is true, thanks for pointing it out. Will change

File: src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs [hunk 4]
```diff
@@ -41,7 +45,11 @@ public async Task<WorkItemScope> BeginWorkItemScopeWhenReadyAsync()
...
+        {
+            await _state.SetStoppedIfStoppingAsync();
```
Comment (dkurepa) [inline src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:51]: Think we'll need one more change here. Since we only set the state to stopped when all consumers finish their current workitem, we'll have to make this place not enter on `Stopping`,  since that's the state the processor will be in while it's waiting for all Consumers to finish

Comment (oleksandr-didyk) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:51]: Can you please expand on what you mean here? Not really sure what you mean here

Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:50]: With your PR, we only call `SetStoppedWhenStoppingAsync` when we finish the last `WorkItem`. But the method that creates work item scope has this `status == Stopped` as it's condition for waiting. So it will keep consuming new WorkItems, and it's possible that one of the listeners will always get work while a different one is processing it, so we'll never be `Stopped`

Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:50]: I think it should just be wait if `status != Working`, it's wrong now too, it will take workitems while it's initializing, I made a bug there too that we just hit

Comment (oleksandr-didyk) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:50]: So if we receive 3 work items, 3 processors would start. Say, 2 finish early and only the last one is processing.

Currently the other two would just continue pulling items from the queue, which is what we would expect, no? And they can't take the item of the 3rd worker that is being processed as its not in the queue anyways.

Or am I misunderstanding you

Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:50]: Yes that is true, but let's say two finish, third one is still working and we do a deployment so the status changes to `Stopping`. In that case we'd wait for someone to call `_state.SetStoppedIfStoppingAsync()` to set the status to `Stopped`. The issue is it's possible for this method to never get called because it's possible `_activeWorkItems` is never `0`. This is because currently, the State check only blocks if the `state = Stopped`. In this case we'd be in `State = Stopping` so we'd keep letting work through, the `_activeWorkItems` would never be 0, so we'd never stop

Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:50]: @oleksandr-didyk I pushed the change to the branch because we need it to fix some stuff, hope that's ok with you

**Review by premun — APPROVED** [review]

File: src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs (hunk 4 above)
Comment (oleksandr-didyk) [reply src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs:50]: > @oleksandr-didyk I pushed the change to the branch because we need it to fix some stuff, hope that's ok with you

Of course it is, no worries. Was out unexpectedly anyways

**Comment by oleksandr-didyk** [comment]
Rolled out by https://github.com/dotnet/arcade-services/releases/tag/v20241016.1-1.1.0-beta.24516.1
